.. raw:: html

    <style>
    #banner_img {
        background: url("_static/design/ivana-cajina-asuyh-_ZX54-unsplash.png") no-repeat center;
        /*Photo by Ivana Cajina on Unsplash*/
        background-size: cover;
    }
    </style>

    <div class="jumbotron jumbotron-fluid" id="banner_img">
    <div class="container-fluid">

:white:`starfish: scalable pipelines for image-based transcriptomics`
=====================================================================

.. raw:: html

    <div style="clear: both"></div>
    <div class="container-fluid">
      <div class="row">
        <div class="col-lg-7">
          <div class="bs-component">
              <a type="button" class="btn btn-primary" href="gallery/quick_start/plot_quick_start.html">quickstart</button></a>
              <a type="button" class="btn btn-primary" href="https://github.com/spacetx/starfish">github</button></a>
              <a type="button" class="btn btn-primary" href="https://forum.image.sc/tag/starfish">user support</button></a>
          </div>
        </div>
      </div>
    </div>
    </div>
    </div>
    </div>
    <br>

    <div class="container-fluid">
    <div class="row">
    <div class="col-md-7">
    <br>

*starfish* is a Python library for processing images of microscopy-based spatial transcriptomics.
It lets you build scalable pipelines that localize and quantify RNA transcripts in image data
generated by any FISH method, from simple RNA single-molecule FISH to combinatorial barcoded
assays. Image processing of an experiment is divided into fields of view (FOV) that correspond to
the data produced by a microscope at a single location on a microscope slide. Each FOV is
processed into a table of spots localized in 3D and then spots are aggregated into a cell x gene
expression matrix by comparing physical positions of spots and cells.

If you want to give it a try, the :ref:`quick start tutorial <quick start>` will guide you from
installation to running a pipeline in under 10 minutes. For more comprehensive instructions on
how to use starfish, see the user guide on :ref:`creating image-based transcriptomics processing
pipelines <creating_an_image_processing_pipeline>`, which organizes and contextualizes the
tutorials on running starfish using the API. Finally, advanced users can examine the
:ref:`Data Structures <data structures>` and :ref:`Help & Reference <help and reference>`
sections to learn more details about starfish and its object models.

In addition to the library of image processing functions, starfish introduces a standardized data
format for microscopy-based spatial transcriptomic assays. Examples and tutorials for formatting
your data into the starfish format can be found in :ref:`starfish data formatting <formatting>`.

.. raw:: html

  </div>
  <div class="col-md-5">
  <h2>Features</h2>

* Formatting Data: :ref:`Tutorials<sphx_glr_gallery_data_formatting>`
* Registering Images: :ref:`API<learn_transform>`
* Filtering Images: :ref:`API<filtering>` | :ref:`Tutorial<section_correcting_images>`
* Finding Spots: :ref:`API<spot_finding>` | :ref:`Tutorial<section_finding_and_decoding>`
* Decoding Spots: :ref:`API<decode_spots>` | :ref:`Tutorial<section_finding_and_decoding>`
* Segmenting Cells: :ref:`API<segmentation>` | :ref:`Tutorial<section_segmenting_cells>`
* Assigning Spots: :ref:`API<target_assignment>` | :ref:`Tutorial<tutorial_assigning_spots>`

.. raw:: html

  </div>
  <div class="col-md-5">
  <h2>Contents</h2>

.. toctree::
   :maxdepth: 1

   getting_started/index
   help_and_reference/index
   usage/index
   processing_at_scale/index
   contributing/index
   creating_an_image_processing_pipeline/index
   gallery/index
   api/index

.. raw:: html

   </div>
   </div>
   </div>
